sim/state.ts

```diff
@@ -0,0 +1,340 @@
+		}
+	}
+
+	private deserializeWithRefs(obj: unknown, battle: Battle) {
+		switch (typeof obj) {
+			case 'undefined':
+			case 'boolean':
+			case 'number':
+				return obj;
+			case 'string':
+				return this.fromRef(obj, battle) || obj;
+			case 'object':
+				if (obj === null) return null;
+				if (Array.isArray(obj)) {
+					const arr = new Array(obj.length);
+					for (const [i, o] of obj.entries()) {
+						arr[i] = this.deserializeWithRefs(o, battle);
+					}
+					return arr;
+				}
+
+				if (this.isActiveMove(obj)) return this.deserializeActiveMove(obj, battle);
+
+				const o: any = {};
+				for (const [key, value] of Object.entries(obj)) {
+					o[key] = this.deserializeWithRefs(value, battle);
+				}
+				return o;
+			case 'function': // lol wtf
+			default:
+				throw new TypeError(`Unexpected typeof === '${typeof obj}': ${obj}`);
+		}
+	}
+
+	private isReferable(obj: object): obj is Referable {
+		// NOTE: see declaration above for why this must be defined lazily.
+		if (!this.REFERABLE) {
+			this.REFERABLE = new Set([
+				Battle, Field, Side, Pokemon, Data.PureEffect,
+				Data.Ability, Data.Item, Data.Move, Data.Template,
+			]);
+		}
+		return this.REFERABLE!.has(obj.constructor);
+	}
+
+	private toRef(obj: Referable): string {
+		// Pokemon's 'id' is not only more verbose than a postion, it also isn't guaranteed
+		// to be uniquely identifying in custom games without Nickname/Species Clause.
+		const id = obj instanceof Pokemon ? `${obj.side.id}${POSITIONS[obj.position]}` : `${obj.id}`;
+		return `[${obj.constructor.name}${id ? ':' : ''}${id}]`;
+	}
+
+	private fromRef(ref: string, battle: Battle): Referable | undefined {
+		// References are sort of fragile - we're mostly just counting on there
+		// being a low chance that some string field in a simulator object will not
+		// 'look' like one. However, it also needs to match one of the Referable
+		// class types to be decode, so we're probably OK. We could make the reference
+		// markers more esoteric with additional sigils etc to avoid collisions, but
+		// we're making a conscious decision to favor readability over robustness.
+		if (ref[0] !== '[' && ref[ref.length - 1] !== ']') return undefined;
+
+		ref = ref.substring(1, ref.length - 1);
+		// There's only one instance of these thus they don't need an id to differentiate.
+		if (ref === 'Battle') return battle;
+		if (ref === 'Field') return battle.field;
+
+		const [type, id] = ref.split(':');
+		switch (type) {
+			case 'Side': return battle.sides[Number(id[1]) - 1];
+			case 'Pokemon': return battle.sides[Number(id[1]) - 1].pokemon[POSITIONS.indexOf(id[2])];
+			case 'Ability': return battle.getAbility(id);
+			case 'Item': return battle.getItem(id);
+			case 'Move': return battle.getMove(id);
+			case 'PureEffect': return battle.getEffect(id);
+			case 'Template': return battle.getTemplate(id);
+			default: return undefined; // maybe we actually got unlucky and its a string
```

Comment:
``switch`` and ``case`` should have the same indentation level

Reply:
Kind of dumb but w/e. Seems like something the linter should enforce?

Reply:
We already enforce that for ``.js`` files. Idk if tslint has that setting.

Reply:
Maybe the new 'eslint-is-tslint' will fix this?

Reply:
Yup!